### ISSUE TITLE
@UnitOfWork not always committing transaction after call

### ISSUE THREAD

**Issue by chris-cynation** [dropwizard/dropwizard#4462]
**Env Details**
Dropwizard version: 2.0.25
Hibernate version: 5.5.8
DB: Postgres 9.6 and 13.4 on AWS RDS.
Java version: 11
Postrgres JDBC driver version: 42.3.1

**Hibernate Entity definition**
	@Id
	@Column(name = "id", updatable = false, nullable = false)
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	private Long id;
	.......
	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (o == null || Hibernate.getClass(this) != Hibernate.getClass(o)) return false;
		Issue issue = (Issue) o;
		return id != null && Objects.equals(id, issue.id);
	}

	@Override
	public int hashCode() {
		return getClass().hashCode();
	}
	
**Details**
I have an App that receives two sequential rest calls.

The first issues a single "delete where xx = 'yy'" on the entity definition above.

The second does processing and on-demand inserts to the same entity.  Since the entity has the @GeneratedValue(strategy = GenerationType.IDENTITY) Hibernate issues an insert directly after the saveOrUpdate call.

The issue, however, is that the delete call sometimes gets issued to the db near the end of the second rest call.

This causes the recent inserts to be deleted.

It doesn't happen every time - which I think is down to how many sets of these calls are being done.

**Main Steps taken to try and resolve**
1. Tried different dropwizard version(initial version had Hibernate 5.5.2) and jdbc driver version.  No difference.
2. Added a flush after the executeUpdate statement.  No effect.

**Workaround found**
If I manually manage the transaction and use @UnitOfWork(transactional = false) and manually call transaction.commit() the problem is fixed.

Is it possibly a configuration issue?

Seems like a bug.  Any thoughts on the issue?

**Comment by zUniQueX** [comment]
The issue you describe sounds mostly like a phantom read problem in the deleting transaction. As the [PostgreSQL documentation](https://www.postgresql.org/docs/9.5/transaction-iso.html) describes, the default isolation level in PostgreSQL is `Read Committed`. With this configuration, phantom reads can occur.

Please try setting the transaction isolation to `Serializable` to exclude an isolation violation. This can be done in the [Dropwizard database configuration](https://www.dropwizard.io/en/latest/manual/configuration.html#database) with the property `defaultTransactionIsolation`.

**Comment by chris-cynation** [comment]
I don't think it's an isolation issue. As there are no read calls.  There's two sql queries (in separate non-concurrent transactions):
delete from
insert into.

**Comment by zUniQueX** [comment]
As you described in the issue definition, the REST calls and therefore the managed transactions interleave. But also when the transactions commit at different times, the database scheduler may schedule them interleaved.

If the insert call is early in the transaction, the tuple gets inserted into the index. When the delete call happens later, it sees the newly inserted entry in the index while identifying the tuples it should delete. As the delete query must find the tuples to delete, it must _read_ the index and therefore can suffer of phantom reads.

I'm not sure, on which side the bug is as interleaved calls are hard to reproduce and test. But I just wanted to make sure it's not a database issue.

**Comment by chris-cynation** [comment]
No, the rest calls are sequential.

Hibernate issues the delete statement to the db too late (seen via hibernate logging showing the delete statement).

Sorry, for the confusion - should have explicitly said hibernate issues the delete too late.

**Comment by zUniQueX** [comment]
Ah, okay. Can you frequently reproduce this, i.e. set up a test?

**Comment by chris-cynation** [comment]
It doesn't happen too frequently in our Env. I'd say it happens like once in every three or so then almost always occurs every time after that first instance.

I can see about trying to create a simple test to isolate further on Mon or so.

**Comment by zUniQueX** [comment]
@chris-cynation I set up a basic test with calls for deleting and inserting entities: https://github.com/zUniQueX/dropwizard/commit/f78643fb158c5a1cb861d288d6076b2bc4464d03. Is this code comparable to your situation?

In this test I cannot reproduce reordering of the db calls with many iterations (here 500). However, if done the requests asynchronously from the client and therefore interleaved, the delete likely happens after the insert call.

**Comment by chris-cynation** [comment]
Hey, 

Apologies, been busy.

I've tried to reproduce but can't as well. I don't think creating it in a similar docker is going to make a difference as I'm running the same dependencies and Java version.

I'll keep an eye out for other instances of the problem.

I've got a workaround where I create a transaction and commit manually.

**Comment by zUniQueX** [comment]
Just to clarify: could it be, that your REST calls get invoked asynchronously in any way?

**Comment by chris-cynation** [comment]
No.

It's possible other rest calls are occurring during that time period that are unrelated. e.g. different endpoint and entity.

The caller of the rest calls is a JS app that does an await.

I can confirm it's synchronous by looking in the logs.

**Comment by chris-cynation** [comment]
I also checked that a return type is specified on the REST endpoint as that can cause issues.

**Comment by zUniQueX** [comment]
Okay, thanks for clarifying this.

This seems to be hard to fix then. Since the tests cannot directly reproduce this situation, I'm not completely sure, that this is a framework issue.

The only difference I can state without further code is that the transaction commit happens later, if the transaction is managed by Dropwizard, as the Jersey processing happens before. But the commit should happen before your next call is triggered, if the calls are sequential.

**Comment by chris-cynation** [comment]
Yeah, I've tried to reproduce it with a complete test setup which is very very similar - but no luck.

I'll close the ticket now.

If I come across the issue again I'll open a new ticket with hopefully the ability to reproduce what's going now.
